### ISSUE TITLE
Handling "No Access" responses

### ISSUE THREAD

**Issue by sqmgh** [ITfoxtec/ITfoxtec.Identity.Saml2#87]
I am trying to deal with an ADFS IdP which is configured to allow only a portion of users to authenticate based on which groups they are apart of.

This, for the most part, works as expected. However, when a user who is not apart of one of the allowed groups tries to authenticate the following exception is thrown: `ITfoxtec.Identity.Saml2.Saml2RequestException: There is not exactly one Assertion element.`

This is true, there are indeed no Assertion elements in the XML. It also technically works as expected; by throwing an exception the user is not allowed in. However, I am looking for a method of more reliably detecting this scenario so I can instead redirect the user to an error page telling the user to go ask for access.

---

The XML returned does include a Status of `urn:oasis:names:tc:SAML:2.0:status:RequestDenied`, and so with my uneducated understanding of the SAML2.0 standard it seems like I could "fix" this by modifying the `Read` function in `Saml2Response.cs` as follows:
```
protected internal override void Read(string xml, bool validateXmlSignature = false)
{
    if (xml.Contains("urn:oasis:names:tc:SAML:2.0:status:RequestDenied"))
    {
        Status = Schemas.Saml2StatusCodes.RequestDenied;
        return;
    }

    base.Read(xml, validateXmlSignature);

    InResponseTo = XmlDocument.DocumentElement.Attributes[Schemas.Saml2Constants.Message.InResponseTo].GetValueOrNull<Saml2Id>();

    Status = Saml2StatusCodeUtil.ToEnum(XmlDocument.DocumentElement[Schemas.Saml2Constants.Message.Status, Schemas.Saml2Constants.ProtocolNamespace.OriginalString][Schemas.Saml2Constants.Message.StatusCode, Schemas.Saml2Constants.ProtocolNamespace.OriginalString].Attributes[Schemas.Saml2Constants.Message.Value].GetValueOrNull<string>());
}
```

Since `XmlDocument` is set in `base.Read(...)`, and that is also where exception happens, I can not just shuffle the lower Status setting line further up in the function. Even if i could, the XML actually contains _two_ Statuses and `Saml2StatusCodeUtil.ToEnum()` ends up setting it to `Saml2StatusCodes.Requester`.

In the end though, this _seems_ to be good enough for my purposes, and even though I don't think it's a good solution, I could technically put in Pull request. However, I am not well versed enough with either SAML2.0 or the library to know if this is opening myself up to security problems (Reading unvalidated xml?), or if there is a more appropriate way of doing this?

Edit: I suppose another question that might answer things is: Is there any other reasons why there wouldn't be exactly one Assertion? Or is it _only_ because the user is does not have access?

**Comment by Revsgaard** [comment]
You need to call the ReadSamlResponse and check the status before you call the Unbind methods which will fall when the XML do not contain one assertion. 

If the status is anything else then success, the Unbind method should not be called.

```
binding.ReadSamlResponse(Request.ToGenericHttpRequest(), saml2AuthnResponse);
if (saml2AuthnResponse.Status != Saml2StatusCodes.Success)
{
      throw new AuthenticationException($"SAML Response status: {saml2AuthnResponse.Status}");
}
binding.Unbind(Request.ToGenericHttpRequest(), saml2AuthnResponse);

```

For support questions please use [Stack Overflow](https://stackoverflow.com/questions/tagged/itfoxtec-identity-saml2). Tag your questions with 'itfoxtec-identity-saml2' and I will answer as soon as possible.

**Comment by sqmgh** [comment]
Ahh, I suppose that does make sense. Looking back I had also done it that way to avoid https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/issues/89. Since it was the only way I could replay detection to work, it didn't occur to me that it might be a bug.

Do you have even the slightest idea of the ETA for when the release of that fix might be? It seems like this has sort of turned into a weird blocker for me.

(Sorry, it seems weird to create a Stack Overflow account, and then start a new question for this...)

**Comment by Revsgaard** [comment]
I'm in doubt which fix you are thinking about?

**Comment by sqmgh** [comment]
Ahh, apologies for not being more clear. 

I had originally removed the call to `ReadSamlResponse()` because I had run into the problem in #89, where tokens were always failing the replay detection. As you say in that issue "There is not security risk in omitting to calling ReadSamlResponse.", and that is what I had done in order to get replay detection working. 

However, it also turns out that I do actually need the `ReadSamlResponse()` in order to (properly) handle the user having no access. So right now I have to choose between replay detection, or wrapping everything in a try catch.

In #89 you mentioned https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/commit/0c0cef05050d633f000be9399dc6a5b1a009a6da would fix the issue, in the next release. So that is the fix I was inquiring about.

---

It appears that while writing this you released a new version. So thank you very much!

**Comment by Revsgaard** [comment]
Okay :)
The fix is part of release [4.7.0](https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/releases/tag/4.7.0).
